Lazy function loading

NathanJPhillips opened the pull request:
This PR is quite long, I'd recommend reviewing commit-by-commit. It has been reviewed in part in Security Scanner but @smowton feels it is ready to be reviewed for inclusion in to CBMC now.

In the end this version eagerly loads all functions (although it does it by pulling them in explicitly). Individual users are free to alter this behaviour.

NathanJPhillips committed:
Allow convert_lazy_method on functions that don't have a lazy body

The call will do nothing if a lazy body doesn't exist to be converted
This allows calling for example on functions that will be stubbed

NathanJPhillips committed:
Add lazy_goto_functions_mapt

Add lazy_goto_functions_mapt, a lazily populated function map, and use it in lazy_goto_modelt
Allow post-processing functions run on a function-by-function basis
Update lazy_goto_modelt to run module passes

NathanJPhillips committed:
Do type-checking as a function pass

After convert each function in convert_lazy_method call instrument and typecheck

NathanJPhillips committed:
Make goto_convert_functions not add directly to function map

This won't be supported with lazy function loading as it won't use goto_functionst

NathanJPhillips committed:
Introduce lazy_goto_modelt

Introduces a class that holds a goto_modelt and language_filest to allow for lazy loading of functions.
Don't eagerly convert functions during initialisation of lazy_goto_modelt. Instead adds a function to allow eager loading at a later stage if required (it always is at the moment).

NathanJPhillips committed:
Moved call to final to lazy_goto_modelt::finalize

Explicitly load any functions added by call to final

NathanJPhillips committed:
Update load_all_functions

Made load_all_functions compatible with some functions having already been loaded by using the lazy load procedure

NathanJPhillips committed:
Encapsulate maps in language_filest

NathanJPhillips committed:
Always allow the main function not to have a body

NathanJPhillips committed:
Recreate initialize in final

Creates the initialize methods again taking account of symbols added to the
symbol table during instantiation of lazy methods since they were first
created.

NathanJPhillips committed:
Don't erase symbol in java_bytecode_convert_methodt::convert

The symbolt was erased and recreated rather than being modified. This was bad as it invalidated the reference to the symbol that we now use later.

NathanJPhillips committed:
Do lowering of java_new as a function-level pass

NathanJPhillips committed:
Fix new tests to use lazy loading

NathanJPhillips committed:
Reset counter used by get_fresh_aux_symbol in unit tests

Otherwise symbols won't have reproducible names

NathanJPhillips commented:
The first 7 commits are from https://github.com/diffblue/cbmc/pull/1558 (waiting to be merged). Do not merge this until that PR has been merged and this one has been rebased on it.

smowton commented:
Tagging do-not-merge until predecessor PRs are merged

smowton commented:
Note: needs @kroening review before merge

smowton commented:
Now mergeable once test failures are addressed

NathanJPhillips committed:
Always convert the main function from bytecode into the symbol_table

Load the main function into the symbol table before generating the entry point so that we have access to its parameter names
Allow attempts to load a function already loaded for non-lazy cases

NathanJPhillips committed:
Encapsulate lazy_goto_modelt::goto_functions

NathanJPhillips commented:
Doxygen check fails because the parser can't cope with lambdas in initialiser lists but there aren't actually any problems with it. All other required CI is passing.

peterschrammel commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -7,27 +7,44 @@
-  goto_functionst &goto_functions;
+  const lazy_goto_functions_mapt<goto_programt> goto_functions;
+
+private:
```
I'd prefer merging respective private (resp. public) sections.

tautschnig replied:
Still applies.

NathanJPhillips replied:
This was discussed offline with Peter and a compromise was reached that grouped some parts together but kept the distinct sections, adding some comment to make them clearer.

tautschnig replied:
Ok; looking at the final version I'm still a bit confused by "Fields" vs "Public fields" not being grouped together/next to each other. (The commit-by-commit review makes this much harder here as apparently later commits added cleanup.)

NathanJPhillips replied:
Peter's concern was that the public API should be in one place. Public fields are, unfortunately, part of that public API and so appear next to public methods after the constructors. Obviously I'd rather not add public fields like this to a new class but I need to maintain compatibility with `goto_modelt` for use in `rebuild_goto_start_function`.

tautschnig replied:
If you want to maintain the separation (space-wise) of `private` fields vs `public` fields then at least the comment should also so `Private fields`. I cannot help but say that the fact that comments are required is [suspicious](http://apdevblog.com/comments-in-code/). But also not a blocker, just voicing (too many?) thoughts.

peterschrammel replied:
I don't see any reason why 
* goto_functions and symbol table can't be next to each other; 
* constructor/assignment/creation can be merged with the methods section
* and the private fields can't be moved to the end of the class declaration.
That would be much less confusing than multiple sections - even with comments.

peterschrammel commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -38,6 +55,12 @@ class lazy_goto_modelt
+    const optionst &options,
+    message_handlert &message_handler);
+
 public:
```
redundant `public`

tautschnig replied:
Still applies.

peterschrammel commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -51,8 +74,86 @@ class lazy_goto_modelt
+      catch(int e)
+      {
+        msg.error()
+          << "process_goto_function: Numeric exception : " << e
```
Side comment: this makes little sense, but I defer the refactoring of this to https://github.com/diffblue/cbmc/issues/1314.

tautschnig replied:
I actually don't get why error handling is done right here. Shouldn't any exceptions just propagate at this point?!

peterschrammel reviewed (changes requested):
Please fix the doxygen issue.
Clang format fixes will be required. 
I am blocking this until the corresponding TG PR passes.

tautschnig commented on src/goto-programs/goto_functions_template.h:
```diff
@@ -49,6 +49,13 @@ class goto_functions_templatet
+  {
+    const auto &it=function_map.find(name);
+    if(it!=function_map.end())
+      function_map.erase(it);
```
`function_map.erase(name);` would do the trick.

tautschnig commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -0,0 +1,72 @@
+
+  /// The model returned here has access to the functions we've already
+  /// loaded but is frozen in the sense that, with regard to the facility to
+  /// load new functions, it has let it go.
```
What has let what go? This comment seems important, but English not being my first language I completely fail to understand it, I'm afraid.

tautschnig replied:
Still applies.

smowton replied:
How about "...in the sense that no further functions can be loaded using this lazy model object."

tautschnig replied:
Is this still open?

tautschnig commented on src/goto-programs/rebuild_goto_start_function.h:
```diff
@@ -21,14 +24,14 @@ class goto_functionst;
   " --function name              set main function name\n"
 
-class rebuild_goto_start_functiont: public messaget
+template<typename goto_modelt>
```
I'd strictly prefer a different parameter name as that is also an existing type name. As such, readers of the code may be confused.

Also applies elsewhere.

NathanJPhillips replied:
I was just trying to reduce code churn by leaving references to the type alone but you're right and I have done this for you.

tautschnig replied:
Thought as much and appreciate that, but also I appreciate your additional changes now :-)

tautschnig commented on src/goto-programs/rebuild_goto_start_function.cpp:
```diff
@@ -107,6 +103,9 @@ void rebuild_goto_start_functiont::remove_existing_entry_point()
 }
+
+template class rebuild_goto_start_function_baset<goto_modelt>;
+template class rebuild_goto_start_function_baset<lazy_goto_modelt>;
```
Why is this explicit instantiation necessary?

NathanJPhillips replied:
Explicit template specialisation is always necessary when a templated classes methods are declared in a cpp file.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -529,10 +531,10 @@ int jbmc_parse_optionst::doit()
 }
 
-bool jbmc_parse_optionst::set_properties()
+bool jbmc_parse_optionst::set_properties(goto_modelt &goto_model)
```
Why is this extra argument being passed when there is no change to body of the function?

NathanJPhillips replied:
`goto_model` was previously a member of this class but now is scoped to where it actually exists.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -571,28 +574,41 @@ int jbmc_parse_optionst::get_goto_program(
+    // and into the caller's goto_model
+    goto_model=lazy_goto_modelt::freeze(std::move(lazy_goto_model));
+    if(goto_model == nullptr)
+      return 6;
```
This would be an early exit without any error message. Help the user understand what's going on, please.

NathanJPhillips replied:
This is only reproducing what would have happened before (passing on the failure of `process_goto_module`). I'd hope that all such paths would already have reported the actual error condition. Would you like this specific message to be followed by a summary (something like "Conversion of lazy goto model to concrete goto model failed. Exiting.")?

tautschnig replied:
Hmm, maybe I'm misunderstanding this here, but I was assuming that the failure of `freeze` is a very peculiar kind of bug. Is it really that a lot more could be going wrong in this step?

NathanJPhillips replied:
Freeze calls `process_goto_functions` (a replacement for the non-function-specific parts of `process_goto_module`) which is where most causes of failure come from (the exception being a failure in the call to `final`, which has its own error message).

tautschnig replied:
Ok, that certainly invalidates my initial concern. But it seems the naming in this API is not particularly intuitive. My preferred search engine tells me that "freeze" means "store (something) at a very low temperature in order to preserve it" or "an act of holding or being held at a fixed level or in a fixed state." Yet you are saying that really quite a lot happens before we might enter such a state of preservation.

NathanJPhillips replied:
Those definitions aren't quite right. To freeze something means to take it from it's previous state and bring it to the state you describe. In this case we are taking a goto_model that doesn't have a fixed set of methods yet and freezing it to create one where methods can no longer be loaded. Taking a read-write object and creating a read-only object from it is a fairly common use of the word freeze in programming.

tautschnig replied:
I had just taken those definitions from Google's Dictionary, so either Google is wrong or I misinterpreted something.

Anyway, I do agree with "taking a read-write object and creating a read-only object" -- but what one does _not_ know is what's going to happen in this process. Function identifiers can be a lot longer than 6 characters, and I'd strongly solicit clarity via descriptive names rather than obfuscation.

smowton replied:
How about simply `run_module_passes_and_freeze`, since the method (a) runs the module-level transform passes and then (b) gives you a plain-old `goto_modelt` back (preventing further function loading), two steps which must take place together since the module-level passes assume function loading is complete.

tautschnig replied:
Sounds good to me!

tautschnig replied:
Could @smowton's suggestion be followed, please?

NathanJPhillips replied:
module means something else in the Java context so after discussion with Chris I've chosen a different name.

tautschnig replied:
Ack.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -626,6 +642,7 @@ int jbmc_parse_optionst::get_goto_program(
 }
 
 bool jbmc_parse_optionst::process_goto_program(
+  goto_modelt &goto_model,
```
Additional parameter without any changes to the function body?

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -790,7 +807,7 @@ bool jbmc_parse_optionst::process_goto_program(
 
 /// invoke main modules
-int jbmc_parse_optionst::do_bmc(bmct &bmc)
+int jbmc_parse_optionst::do_bmc(bmct &bmc, goto_modelt &goto_model)
```
Additional parameter without any changes to the function body?

tautschnig commented on src/goto-programs/goto_convert_functions.h:
```diff
@@ -16,6 +16,9 @@ Date: June 2003
 #include "goto_convert_class.h"
+#include "goto_functions.h"
+
+class goto_modelt;
```
This forward declaration seems unnecessary given the `#include "goto_model.h"` above.

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -0,0 +1,170 @@
+#ifndef CPROVER_GOTO_PROGRAMS_LAZY_GOTO_FUNCTIONS_MAP_H
+#define CPROVER_GOTO_PROGRAMS_LAZY_GOTO_FUNCTIONS_MAP_H
+
+#include <type_traits>
```
Are all the typedefs introduced actually necessary? There seems to be a lot of extra work going on just to make those typedefs possible.

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -128,6 +128,13 @@ class lazy_goto_functions_mapt final
+  {
+    const auto &it=goto_functions.find(name);
+    if(it!=goto_functions.end())
+      goto_functions.erase(it);
```
`goto_functions.erase(name);` will do.

tautschnig commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -7,33 +7,52 @@
 #define CPROVER_GOTO_PROGRAMS_LAZY_GOTO_MODEL_H
 
 #include <util/language_file.h>
+#include <util/exit_codes.h>
```
This feels VERY wrong.

NathanJPhillips replied:
Copied from Daniel's recent code in `initialize_goto_model`.

tautschnig replied:
That code doesn't actually do exits, does it? At least not that I could find it.

NathanJPhillips replied:
Sorry, I got the source wrong. It's actually from here: https://github.com/diffblue/cbmc/commit/961d33aff5c6253ae3695c27fdf3d28c213ee65f

tautschnig replied:
Yes, but that include is _not_ used in `goto-programs`. It should be placed within command-line front-ends. Doing a hard exit somewhere deep in the code feels wrong.

NathanJPhillips replied:
This was an attempt to remove duplicated error handling out of each driver program. Should I put it back into each place that it was moved here from? Do @peterschrammel or @kroening have views on this?

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -755,7 +757,7 @@ bool jbmc_parse_optionst::process_goto_program(
            goto_model,
            get_message_handler()))
-        return true;
+        return false;
```
What's going on here?

tautschnig commented on src/goto-programs/Makefile:
```diff
@@ -27,6 +27,7 @@ SRC = basic_blocks.cpp \
       interpreter.cpp \
       interpreter_evaluate.cpp \
       json_goto_trace.cpp \
+      lazy_goto_model.cpp \
```
This isn't in the right commit, which will make `git bisect` harder.

NathanJPhillips replied:
Well caught. This was originally the commit where the cpp file was introduced but I pushed it back to an earlier commit without updating the Makefile. Mea culpa.

tautschnig replied:
(Just for the record: this has been fixed.)

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <util/config.h>
 #include <util/language.h>
 #include <util/unicode.h>
+#include <util/journalling_symbol_table.h>
```
Nit pick: lexicographic sorting of includes.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -207,6 +208,24 @@ void lazy_goto_modelt::load_all_functions() const
+    journalling_symbol_tablet::wrap(this->symbol_table);
+  if(language_files.final(symbol_table))
+  {
+    msg.error() << "FINAL STAGE CONVERSION ERROR" << messaget::eom;
```
What would a user know about "FINAL STAGE"? (Actually even as a developer I don't know what that is.)

NathanJPhillips replied:
Agreed. This has been copied from initialize_goto_model. Someone suitably qualified to know what the message should be should fix it there and I'll copy the fix in here if that merges first.

tautschnig replied:
That was 98643d7335 - @kroening ?!

tautschnig replied:
Still calling upon @kroening to clarify.

peterschrammel replied:
I don't think the user knows about 'FINAL STAGE'. 'CONVERSION ERROR' alone is sufficient (even though I can imagine better error messages than 'CONVERSION ERROR' -- we can address this when fixing up the error handling. Then an appropriate error message can be produced at the location where the 'ERROR' happens.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -207,6 +208,24 @@ void lazy_goto_modelt::load_all_functions() const
+  if(language_files.final(symbol_table))
+  {
+    msg.error() << "FINAL STAGE CONVERSION ERROR" << messaget::eom;
+    return false;
```
This is not the usual convention in this code base: return `true` in case of errors.

NathanJPhillips replied:
The reality is that both are used quite a lot in the code base. This is not part of the API of this class so I'll leave it unless you really care.

tautschnig replied:
Any examples of cases where `false` is used to signify an error? (Yes, I do really care.)

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -207,6 +208,24 @@ void lazy_goto_modelt::load_all_functions() const
+    {
+      // Re-convert any that already exist
+      goto_functions.unload(updated_symbol_id);
+      goto_functions.at(updated_symbol_id);
```
I'm not sure I enjoy the implicitness here.

NathanJPhillips replied:
I see what you mean. I think in any "real" policy (not just "load all") the body of this function would be explored, making it more explicit.

tautschnig replied:
Would you mind adding some API (unless it exists?) so that you can do `goto_functions.convert(updated_symbol_id);` instead of using `at`? I'm perfectly fine with that function just wrapping a call to `at`, but this code is not very descriptive (as witnessed by the fact that you felt compelled to add a comment).

tautschnig replied:
Is this acted upon in some later commit?

tautschnig commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -98,6 +98,9 @@ class lazy_goto_modelt
   // Helpers
 private:
+  static bool is_function_symbol(const symbolt &symbol)
+  { return !symbol.is_type && !symbol.is_macro && symbol.type.id()==ID_code; }
```
This is actually useful elsewhere. Maybe move it to symbol.h?

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -202,13 +202,37 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
 {
-  goto_convert(*goto_model, message_handler);
-}
+  std::vector<irep_idt> fn_ids_to_convert;
```
Why use a `std::vector` here?

NathanJPhillips replied:
I think you might be suggesting that this is declared in the wrong scope, in which case I agree and have fixed it. If you're saying it's the wrong datatype then you'll have to suggest a more suitable alternative because it's not obvious to me.

tautschnig replied:
When I wrote that comment I was thinking about the data type (but, yes, the placement was another issue). You are building a queue of unknown size, hence a `std::list` or `std::deque` would seem much more appropriate. What were your reasons for choosing `std::vector` instead?

smowton replied:
I think `std::vector` is best here. Except for large list items which are expensive to copy, if you don't need `list`'s iterator stability over mutation or `deque`'s efficient push-front operation (which we don't here, since the queue is built once and then drained fully) then `vector` is the highest performance STL sequence container.

tautschnig replied:
There are  two reasons I was surprised to see `std::vector` here: 1) If performance were the reason to use a `std::vector`, then I would have expected some use of `reserve` to avoid repeated re-allocation. 2) If performance wasn't the reason, then it must be another unique capability, such as random access, but that isn't used either. Thus a data structure most naturally fitting the demands of a queue of unknown size should be used, IMHO.

smowton replied:
We could do (1) but we'd need to iterate over the symbol table twice, probably not a saving (or vastly over-reserve, e.g. to symbol_table.size()). Deque and list are both more costly (somewhat and hugely, respectively; http://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html). Deque and list offer no advantages for this use case. Therefore we default to `vector`, which is pretty cheap even with re-allocation considering there's only at worst one realloc copy per push-back [for an exponential reallocation policy; Microsoft's quadratic policy has worst case log2(size) copies per push-back]

tautschnig replied:
Ok, thanks a lot for the pointer!

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -202,13 +202,37 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
+    table_size=new_table_size;
+
+    // Find symbols that correspond to functions
+    fn_ids_to_convert.clear();
```
Just use an object that's local to the loop body, please.

NathanJPhillips replied:
This is the scoping I mentioned in my last comment. Hopefully that's all you meant.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -202,13 +202,37 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
+    // Repeat while new symbols are being added in case any of those are
+    // stubbed functions. Even stubs can create new stubs while being
+    // converted if they are special stubs (e.g. string functions)
+    new_table_size=symbol_table.symbols.size();
```
Can't you use the journalling symbol table here?

NathanJPhillips replied:
That would be nice, but unfortunately not as the thing adding symbols is using a cached reference to the global symbol table.

tautschnig commented on src/java_bytecode/java_entry_point.cpp:
```diff
@@ -485,6 +485,49 @@ bool java_entry_point(
+  messaget message(message_handler);
+  main_function_resultt res=
+    get_main_symbol(symbol_table, main_class, message_handler);
+  if(res.status!=main_function_resultt::Success)
```
Not your bug, but I'm surprised that CamelCase made it into the code base.

NathanJPhillips replied:
Can't see the line you're referring to.

tautschnig replied:
I'm referring to the use of `Success`, where I would have expected `SUCCESS`.

tautschnig replied:
Hmm, 0aeb459def actually is yours ... Please fix.

NathanJPhillips replied:
This will be a separate PR after this one is merged.

tautschnig replied:
Thanks!

tautschnig commented on src/util/language_file.h:
```diff
@@ -104,7 +99,8 @@ class language_filest:public messaget
-    if(has_lazy_method(id))
+    lazy_method_mapt::iterator it=lazy_method_map.find(id);
+    if(it!=lazy_method_map.end())
       lazy_method_map.at(id)->convert_lazy_method(id, symbol_table);
```
Should use `it` instead of doing another lookup via `lazy_method_map.at`.

tautschnig commented on src/goto-programs/remove_java_new.cpp:
```diff
@@ -0,0 +1,103 @@
+    side_effect_exprt malloc_expr(ID_allocate);
+    malloc_expr.copy_to_operands(object_size);
+    // could use true and get rid of the code below
+    malloc_expr.copy_to_operands(false_exprt());
```
Somewhat unrelated: it may be time to do what the comment above says, at which point I'm really wondering whether there is a place for `java_new` at all?

This would likely also yield performance improvements as constant propagation on structs would start happening.

tautschnig commented on regression/goto-cc-cbmc/check-fps-removed/test.desc:
```diff
@@ -0,0 +1,5 @@
+main.c
+--show-goto-functions
+^EXIT=0$
+^SIGNAL=0$
```
What is this actually testing for?

NathanJPhillips replied:
As per commit message:
These tests ensure per-function and module passes are run on functions loaded from goto binaries as they are not run by goto-cc

tautschnig replied:
How does that particular test ensure this? AFAIK `--show-goto-functions` would work just fine in presence of function pointers etc. (I can see that the other test does what it's supposed to do.)

NathanJPhillips replied:
This one in particular checks that function pointers are removed (if they're not then show-goto-functions will cause an error)

tautschnig replied:
Why would show-goto-functions require function pointer removal? That would be a bug in show-goto-functions.

NathanJPhillips replied:
Just realised that since Daniel asked us to back lazy loading out of CBMC these tests no longer relate to this PR. I'll take them out and fix them up for a separate PR.

NathanJPhillips committed:
Change template parameter name to not clash with existing type

NathanJPhillips committed:
Fix new test to run remove_java_new pass

NathanJPhillips committed:
Update test to handle changed symbol creation order

Remove line that depends on generated symbol name - Lucas confirms it wasn't the purpose of the test anyway.

NathanJPhillips committed:
Store symbol type instead of followed type when clean casting

Don't store followed type as type of superclass when doing clean cast as it may later change since we can now convert more methods after typechecking and doing this may change the type

NathanJPhillips commented:
All tests now pass both here and in test-gen. Could @peterschrammel, @kroening or @tautschnig have another look and give final approval for both PR's (this and https://github.com/diffblue/test-gen/pull/1236)?

tautschnig commented on src/goto-programs/goto_function_template.h:
```diff
@@ -0,0 +1,87 @@
+/// A converted function, including its body.
+
+#ifndef CPROVER_GOTO_PROGRAMS_GOTO_FUNCTION_TEMPLATE_H
+#define CPROVER_GOTO_PROGRAMS_GOTO_FUNCTION_TEMPLATE_H
```
What is the advantage of splitting this out?

NathanJPhillips replied:
Good question. I think not much any more. This PR has unfortunately had major changes which mean that lots of things have been backed out. Would you prefer that I undid this change or moved it into a separate PR?

tautschnig replied:
I think you could just drop this one commit to the PR. (If you think it is useful elsewhere then just make sure you keep the commit around.)

tautschnig commented on src/util/language_file.h:
```diff
@@ -92,19 +91,17 @@ class language_filest:public messaget
     symbol_tablet &symbol_table)
   {
-    return lazy_method_map.at(id)->convert_lazy_method(id, symbol_table);
+    PRECONDITION(symbol_table.symbols.count(id) != 0);
```
Use `symbol_table.has_symbol(id)`.

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -0,0 +1,162 @@
+#include <unordered_set>
+#include "goto_functions.h"
+#include "goto_convert_functions.h"
+#include <util/type_traits.h>
```
I don't think this is needed/used anymore.

tautschnig commented on src/util/type_traits.h:
```diff
@@ -0,0 +1,71 @@
+/// Useful functions for manipulating/inspecting types.
+
+#ifndef CPROVER_UTIL_TYPE_TRAITS_H
+#define CPROVER_UTIL_TYPE_TRAITS_H
```
Please remove this file.

tautschnig commented on src/goto-programs/remove_exceptions.cpp:
```diff
@@ -551,39 +559,52 @@ void remove_exceptionst::instrument_exceptions(
+    INVARIANT(
+      false,
+      "Can't currently only remove added instanceof, use "
+        "also_remove_instanceof");
```
Please move the condition into the `INVARIANT`.

Actually I think this value of `type` should not exist: if it isn't supported, then make that explicit via the API. Changing that later isn't such a big pain. Having an API part of which is dead is.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -695,10 +697,9 @@ bool jbmc_parse_optionst::process_goto_functions(
-    remove_instanceof(goto_model);
+    // remove catch and throw (introduces instanceof but request it is removed)
+    remove_exceptions(
+      goto_model, remove_exceptions_typest::ALSO_REMOVE_INSTANCEOF);
```
What is the value of a separate call to `remove_instanceof`?

tautschnig commented on src/goto-programs/remove_exceptions.h:
```diff
@@ -21,7 +21,16 @@ Date:   December 2016
+  DONT_REMOVE_INSTANCEOF,
+  REMOVE_ADDED_INSTANCEOF,
+  ALSO_REMOVE_INSTANCEOF,
+};
```
I find this pretty confusing. (And it's also undocumented.)

tautschnig commented on src/goto-programs/remove_instanceof.h:
```diff
@@ -17,8 +17,12 @@ Author: Chris Smowton, chris.smowton@diffblue.com
 void remove_instanceof(
-  symbol_tablet &symbol_table,
-  goto_functionst &goto_functions);
+  goto_functionst::goto_functiont &function,
```
Pass a `goto_programt` instead.

tautschnig replied:
Thoughts?

NathanJPhillips replied:
Would you mind explaining your thinking behind this suggestion? No particular objections but don't want to make changes for no reason.

tautschnig replied:
You are only using a single member of the `goto_functiont` class (the `body`), hence this function seems applicable in all contexts where the potential caller holds a `goto_programt`. It thus makes the function more general. (It may be worth scrutinising the code base for all uses of `goto_functiont` as really only those cases where type information is required should be using that.

NathanJPhillips replied:
A quick search reveals the following methods that look to me like passes using `goto_functiont`:
`remove_complex`, `remove_vector`, `remove_java_new`, `adjust_float_expressions`, `rewrite_union`, `remove_asm`
Making `remove_instanceof` not use it when these do seems like a bad idea. If you want to change them all then feel free to create an issue.

peterschrammel replied:
As far as I am aware the code base only passes goto_programt for manipulations of goto programs.

tautschnig replied:
... which is what all these functions do. Added as #1681.

tautschnig commented on src/goto-symex/adjust_float_expressions.cpp:
```diff
@@ -200,6 +207,6 @@ void adjust_float_expressions(
-  namespacet ns(goto_model.symbol_table);
-  adjust_float_expressions(goto_model.goto_functions, ns);
+  adjust_float_expressions(
+    goto_model.goto_functions, namespacet(goto_model.symbol_table));
```
Looks like unnecessary code churn.

tautschnig commented on src/goto-symex/rewrite_union.cpp:
```diff
@@ -128,6 +135,6 @@ void rewrite_union(
-  namespacet ns(goto_model.symbol_table);
-  rewrite_union(goto_model.goto_functions, ns);
+  rewrite_union(
+    goto_model.goto_functions, namespacet(goto_model.symbol_table));
```
As above: unnecessary code churn.

tautschnig commented on src/java_bytecode/java_bytecode_typecheck.cpp:
```diff
@@ -38,34 +38,35 @@ void java_bytecode_typecheckt::typecheck()
-
-  std::vector<irep_idt> identifiers;
+  journalling_symbol_tablet::changesett identifiers;
   identifiers.reserve(symbol_table.symbols.size());
```
That's no longer necessary.

NathanJPhillips replied:
Do you mean the call to reserve is no longer necessary because it's not a vector? http://en.cppreference.com/w/cpp/container/unordered_set/reserve

tautschnig replied:
Yes, because it's not a vector anymore and also it's an empty set. Thus it has no effect as far as I am aware.

tautschnig replied:
Thoughts?

NathanJPhillips replied:
Does the link in my previous comment address your concerns?

tautschnig replied:
Not really, because I'd claim this operation has no effect on an empty set?!

NathanJPhillips replied:
I think an empty set has buckets allocated ready to receive elements. This will increase the number of buckets allocated now if that number would otherwise have been increased later.

tautschnig replied:
True, it's only the `rehash` part that has no effect. My apologies for the noise.

tautschnig reviewed (commented):
I have only reviewed the first 8 commits for now, but gathered sufficiently many change requests I'd say.

It's probably redundant to say, but this is far from an optimal pull request: there is a lot of refactoring that is applicable even without lazy function loading. As such reviewing requires a major time commitment. As finding such time slots is hard, reviewing will have to happen in chunks.

NathanJPhillips committed:
Allow post-processing functions to run on a function-by-function basis

NathanJPhillips commented:
Made changes to address 12 of @tautschnig 's comments and replied to the others.

smowton commented on src/java_bytecode/java_entry_point.cpp:
```diff
@@ -485,6 +485,49 @@ bool java_entry_point(
 }
 
+/// Creates the initialize methods again taking account of symbols added to the
+/// symbol table during instantionation of lazy methods since they were first
```
sp instantiation

smowton commented on unit/pointer-analysis/custom_value_set_analysis.cpp:
```diff
@@ -183,6 +184,9 @@ SCENARIO("test_value_set_analysis",
 
     namespacet ns(goto_model.symbol_table);
 
+    // VSA doesn't currently support java_new as an allocater
```
sp allocator

smowton reviewed (commented):
Some small nitpicks

smowton commented:
@tautschnig are you still reviewing?

tautschnig commented:
> @tautschnig are you still reviewing?

Returning to this now. (It's not all that easy to allocate large chunks of time...)

peterschrammel commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -196,13 +196,36 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
 bool lazy_goto_modelt::finalize()
 {
-  messaget msg(language_files.get_message_handler());
+  messaget msg(message_handler);
```
That change is unrelated to this commit. Could be squashed into the previous commit.

peterschrammel commented on src/goto-programs/class_identifier.cpp:
```diff
@@ -71,3 +71,29 @@ exprt get_class_identifier_field(
   return build_class_identifier(deref, ns);
 }
+
+void set_class_identifier(
```
Please add some doxygen

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -0,0 +1,179 @@
+        throw 0;
+      }
+
+      std::pair<language_filest::file_mapt::iterator, bool> result =
```
Nit picking: space around `=` or no space - please just don't mix it within a single commit.

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -0,0 +1,161 @@
+private:
+  typedef std::map<key_type, goto_function_templatet<bodyt>> underlying_mapt;
+  underlying_mapt &goto_functions;
+  mutable std::unordered_set<irep_idt, irep_id_hash> processed_functions;
```
Why is this mutable?

NathanJPhillips replied:
Because its internal state that can be modified without changing the external interface of the class.

tautschnig replied:
Are there particularly good reasons for making that distinction, or reasons for marking methods `const` when they actually change the internal state? I'd say that `mutable` should only be used when a member is not part of the object's state (such as a cache).

NathanJPhillips replied:
I agree with your definition of when `mutable` should be used. To me this class is a map from method names to `programt` function bodies and which ones are loaded in to the bytecode/`codet`/`programt` formats at any time is the "cache". `processed_functions` stores which functions are already fully available in the `programt` state.

tautschnig replied:
Ok, makes sense. Yet seemingly I wasn't able to infer this from the name, so maybe you could add the above explanation as a comment?

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -0,0 +1,161 @@
+  // This is mutable because it has internal state that it changes during the
+  // course of conversion. Strictly it should make that state mutable or
+  // recreate it for each conversion, but it's easier just to store it mutable.
+  mutable goto_convert_functionst convert_functions;
```
Wouldn't this be a good time to fix it then?

NathanJPhillips replied:
After trying to get this through for months this seems like a bad time to put it on hold again while I try and get another PR through to make this change.

tautschnig replied:
I accept that my request may be beyond this PR, and creating an issue to track the need for this change is a solution that would be perfectly acceptable to me. (But I wasn't impressed by your choice of words as it wasn't me who put up a pull request with 23 commits and a diff of +1,124 −328.)

NathanJPhillips replied:
https://github.com/diffblue/cbmc/issues/1675

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -0,0 +1,161 @@
+private:
+  // This returns a non-const reference, but you should const it before
+  // returning it to clients from a const method
+  reference ensure_function_loaded(const key_type &name) const
```
The above "contract" seems breached by `at`?! Either the comment or the implementation is wrong.

NathanJPhillips replied:
The return type of the const `at` implementation is `const_mapped_type`

tautschnig replied:
I was thinking of the non-const `at` implementation. Maybe I'm misunderstanding the comment.

NathanJPhillips replied:
I'll reword the comment.

tautschnig commented on src/util/symbol.h:
```diff
@@ -109,6 +109,9 @@ class symbolt
   }
+
+  bool is_function() const
+  { return !is_type && !is_macro && type.id()==ID_code; }
```
Start the `return` on a new line, please.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -196,13 +196,36 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
 
+    // Access all functions to convert them
+    for(const irep_idt &symbol_name : fn_ids_to_convert)
+      goto_functions.at(symbol_name);
```
The above comment should not be necessary: please use a function name that is descriptive.

tautschnig commented on src/util/language_file.h:
```diff
@@ -59,21 +59,36 @@ class language_filet final
+  language_filet &add_file(const std::string &filename)
+  {
+    language_filet language_file;
+    language_file.filename=filename;
```
While doing the refactoring, would you mind replacing the default constructor of `language_filet` by `explicit language_filet(const std::string &filename):filename(filename) {}`?

tautschnig commented on src/util/language_file.h:
```diff
@@ -59,21 +59,36 @@ class language_filet final
+  void remove_file(const std::string &filename)
+  {
+    // TODO: Clear relevant entries from lazy_method_map
+    // where second == &file_map.at(filename)
```
Why is this not addressed right away?

tautschnig commented on src/java_bytecode/java_entry_point.cpp:
```diff
@@ -485,6 +485,49 @@ bool java_entry_point(
 
+/// Creates the initialize methods again taking account of symbols added to the
+/// symbol table during instantiation of lazy methods since they were first
+/// created,
```
Just putting it here because github does not permit comments on commit messages: the commit message lacks an answer to the "Why?" question. I think (but am not sure) that the above comment might be the desired answer.

tautschnig commented on src/java_bytecode/java_entry_point.cpp:
```diff
@@ -485,6 +485,49 @@ bool java_entry_point(
+    // it now
+    return res.status==main_function_resultt::Error;
+  }
+  symbolt symbol=res.main_function;
```
Why is this not a const reference? Yet really it seems all you need is the location.

NathanJPhillips replied:
I'm not sure this is stable over `create_initialize`. If you can guarantee that it is then we could grab it at this point.

tautschnig replied:
I'd say those symbols should be stable, but better safe than sorry (as you'd be relying on implementation internals). So what I'd recommend is only taking a copy of the location rather than the full symbol, as this is all that you use.

NathanJPhillips replied:
I'm worried that there may be a location numbering update when a new function is loaded but to be honest I don't really understand how these things work so maybe the worry is unfounded.

tautschnig replied:
In the final version of this series, line 502 is the only place where `symbol` is referred to, and then only `symbol.location` is used. Thus my suggestion to store that `source_locationt` rather than the full `symbolt`.

smowton replied:
`res` already contains a `symbolt` (i.e. a value type, not a reference), so I suspect it makes no difference whether we take an additional copy here or not

NathanJPhillips replied:
Having looked at `create_initialize` it looks OK so I've moved the whole access later. If the tests pass we'll assume it's fine!

tautschnig commented on src/goto-programs/remove_java_new.h:
```diff
@@ -0,0 +1,27 @@
+#define CPROVER_GOTO_PROGRAMS_REMOVE_JAVA_NEW_H
+
+#include "goto_model.h"
+#include <util/message.h>
```
A forward declaration would do.

tautschnig commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -295,6 +295,10 @@ bool java_bytecode_languaget::generate_support_functions(
 
+  // Load the main function into the symbol table to get access to its
+  // parameter names
+  convert_lazy_method(res.main_function.name, symbol_table);
```
I'm afraid I don't understand the commit subject. None of the changes here seem to be doing anything in `main`.

NathanJPhillips replied:
The line you're commenting on says `main_function` in the middle. That's the one we're faulting in.

tautschnig replied:
As English isn't my first language: what does "faulting" mean here?

NathanJPhillips replied:
Sorry for the confusion. Think of it like a page fault in a memory mapper. We're converting the main function and adding it to the cache of loaded functions.

tautschnig replied:
Ok, I was guessing along those lines, but if you can find a way of wording the commit message so that even simple-minded people like myself can grasp it this would be appreciated.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -205,9 +205,8 @@ void lazy_goto_modelt::load_all_functions() const
-    // Access all functions to convert them
     for(const irep_idt &symbol_name : fn_ids_to_convert)
-      goto_functions.at(symbol_name);
+      goto_functions.ensure_function_loaded(symbol_name);
```
Thanks a lot for this; it would be great if that change could be folded into the code introducing the line the first place so as to avoid redundant comment...

tautschnig reviewed (commented):
I have now gone over all commits (again). More comments and some reminders of past ones that never got to see a response.

peterschrammel commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -625,14 +641,44 @@ int jbmc_parse_optionst::get_goto_program(
-bool jbmc_parse_optionst::process_goto_program(
-  const optionst &options)
+void jbmc_parse_optionst::process_goto_function(
+  const irep_idt &function_name,
```
`function_name` is unused

NathanJPhillips replied:
It is in this implementation, but may be required by implementations that do other function-level passes that need access to function-related symbols.

tautschnig replied:
Are the actual uses thereof? Else it could easily be changed later on as the need arises, I'd say.

peterschrammel replied:
'may be required' -- I don't think we should do something because it may be required. Anyways, this refactoring is outside the scope of this PR.

NathanJPhillips replied:
There were passes that needed this. They are now not included in this PR.

tautschnig replied:
I'd be keen to see those as I'm not at all convinced they should be getting the function name in this way. I'd have recommended `get_function_id()`.

NathanJPhillips replied:
If you're saying that we can get the function name from the `functiont` then this parameter isn't needed. Shall I remove it?

tautschnig replied:
You can get it from a non-empty `goto_programt`. If that's all you need (and once #1677 is merged) then you should be fine without that parameter.

NathanJPhillips replied:
OK, removed.

peterschrammel commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -625,14 +641,44 @@ int jbmc_parse_optionst::get_goto_program(
-  const optionst &options)
+void jbmc_parse_optionst::process_goto_function(
+  const irep_idt &function_name,
+  goto_functionst::goto_functiont &function,
```
should pass `goto_programt`

NathanJPhillips replied:
See discussion above about various passes that operate on `goto_functiont`. This change should be made as part of any PR that addresses those instances.

peterschrammel commented on src/goto-programs/remove_asm.cpp:
```diff
@@ -305,24 +298,24 @@ void remove_asmt::process_function(
 /// removes assembler
-void remove_asmt::operator()()
+void remove_asm(
+  goto_functionst::goto_functiont &goto_function,
```
I don't see `goto_function.first` being used anywhere here.

NathanJPhillips replied:
I don't understand the comment. This argument is just passed to `remove_asmt::process_function`, which Daniel wrote.

tautschnig replied:
I'd propose to leave this to #1681.

NathanJPhillips commented:
Missed the forward declaration comment but addressed all other points. I'll let the CI run for now and push the one I missed with the next set of requested changes once I've heard back from @tautschnig and @peterschrammel on the other points.

NathanJPhillips commented:
I think everything is answered, fixed or an issue created to track it now. Tests are all passing.

peterschrammel commented:
TG passes too. Seems ready to go then.

peterschrammel commented:
I am sticking a 'do not merge' label on it to reduce risk for a planned TG release on 3rd Jan. We'll merge it right after that. 

Besides, an approval is required by @tautschnig or @kroening.

tautschnig commented:
Added a new label (Merge on 4 Jan) as there might be more coming up the next days that shouldn't get merged until then. Obviously I know nothing about about the release so feel free to change that.
